polygeist/test/polygeist-opt/sycl/nd-range-analysis.mlir

```diff
@@ -0,0 +1,178 @@
+  %global_size = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::range", (struct<"class.sycl::_V1::detail::array", (array<1 x i64>)>)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
+  %local_size = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::range", (struct<"class.sycl::_V1::detail::array", (array<1 x i64>)>)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
+  %offset = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::id", (struct<"class.sycl::_V1::detail::array", (array<1 x i64>)>)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
+  sycl.host.constructor(%local_size, %c21) {type = !sycl_range_1_} : (!llvm.ptr, i64) -> ()
+  scf.if %arg0 {
+    sycl.host.constructor(%global_size, %c42) {type = !sycl_range_1_} : (!llvm.ptr, i64) -> ()
+    sycl.host.constructor(%offset, %c10) {type = !sycl_id_1_} : (!llvm.ptr, i64) -> ()
+    sycl.host.constructor(%ndr, %global_size, %local_size, %offset) {type = !sycl_nd_range_1_} : (!llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> ()
+  } else {
+    sycl.host.constructor(%global_size, %c10) {type = !sycl_range_1_} : (!llvm.ptr, i64) -> ()
+    sycl.host.constructor(%offset, %c10) {type = !sycl_id_1_} : (!llvm.ptr, i64) -> ()
+    sycl.host.constructor(%ndr, %global_size, %local_size, %offset) {type = !sycl_nd_range_1_} : (!llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> ()
+  }
+  %1 = llvm.load %ndr {tag = "join_ndr"} : !llvm.ptr -> i32
+  llvm.return %ndr : !llvm.ptr
+}
+
+// CHECK-LABEL: test_tag: join_ndr_top:
+// CHECK-NEXT:    operand #0
+// CHECK-NEXT:    nd_range:
+// CHECK-NEXT:      <unknown>
+llvm.func @join_ndr_top(%arg0: i1) -> !llvm.ptr {
+  %c1 = arith.constant 1 : i32
+  %c10 = arith.constant 10 : i64
+  %c21 = arith.constant 21 : i64
+  %c42 = arith.constant 42 : i64
+  %ndr = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::nd_range", (struct<"class.sycl::_V1::range", (struct<"class.sycl::_V1::detail::array", (array<1 x i64>)>)>)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
+  %global_size = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::range", (struct<"class.sycl::_V1::detail::array", (array<1 x i64>)>)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
+  %local_size = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::range", (struct<"class.sycl::_V1::detail::array", (array<1 x i64>)>)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
+  %offset = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::id", (struct<"class.sycl::_V1::detail::array", (array<1 x i64>)>)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
+  scf.if %arg0 {
+    sycl.host.constructor(%global_size, %c42, %c42) {type = !sycl_range_2_} : (!llvm.ptr, i64, i64) -> ()
+    sycl.host.constructor(%local_size, %c21, %c21) {type = !sycl_range_2_} : (!llvm.ptr, i64, i64) -> ()
+    sycl.host.constructor(%offset, %c10, %c10) {type = !sycl_id_2_} : (!llvm.ptr, i64, i64) -> ()
+    sycl.host.constructor(%ndr, %global_size, %local_size, %offset) {type = !sycl_nd_range_2_} : (!llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> ()
```

Comment:
should element type of %ndr matches the sycl.host.constructor type attribute? or it is allowed to be different?

Reply:
`%ndr` does not have a type, as it is an opaque pointer. I'd argue we should not look at the `llvm.alloca` element type as we could potentially have flows, e.g., involving `reinterpret_cast` in which something similar to what we're testing here happens.